### PR TITLE
Extract CPU log_softmax kernels to header

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #156243

This allows sharing them with ExecuTorch.

Differential Revision: [D76830114](https://our.internmc.facebook.com/intern/diff/D76830114/)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

### CODE REVIEW

File: aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+    int64_t k = i % num_chunks;
+    int64_t inner_idx_begin = k * chunk_size;
+    int64_t size = std::min(chunk_size, inner_size - inner_idx_begin);
+
+    fVec zero_fvec = fVec(float(0));
+    fVec min_fvec = fVec(-std::numeric_limits<float>::infinity());
+    int64_t d0 = 0;
+    for (; d0 < size - (size % Vec::size()); d0 += Vec::size()) {
+      min_fvec.store(input_max_data + d0);
+      min_fvec.store(input_max_data + d0 + fVec::size());
```
Comment (janeyx99) [inline aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h:247]: This looks added from the original implementation (and I see this throughout the below) -- are we doing 2x as much vectorization? Or...?

File: aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        float data_val = float(input_ptr[d1]);
+        float max_val = input_max_data[d1];
+        input_max_data[d1] = data_val > max_val ? data_val : max_val;
+        input_buffer_ptr[d1] = data_val;
+      }
+    }
+
+    // compute sum of (x - max).exp()
+    for (int64_t dim_idx = 0; dim_idx < dim_size; dim_idx++) {
+      float* input_buffer_ptr = input_buffer_data + dim_idx * chunk_size;
```
Comment (janeyx99) [inline aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h:287]: how come float* instead of const scalar_t*?

**Review by janeyx99 — COMMENTED** [review]
CI will check correctness--is there a plan to check perf?

File: aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h (hunk 1 above)
Comment (swolchok) [reply aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h:247]: it's not new; there are multiple kernels being moved (there's an enable_if). you can verify moved code in Phabricator

File: aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h (hunk 2 above)
Comment (swolchok) [reply aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h:287]: not new

**Review by janeyx99 — APPROVED** [review]
Oooh

**Comment by swolchok** [comment]
hmm, these linux-aarch64 failures look nonsensical (test_forward_nn_CrossEntropyLoss_cpu_float64 is failing). rebasing.

**Comment by swolchok** [comment]
hm, second failure does mention log_softmax, perhaps it is related

**Comment by swolchok** [comment]
failures *are* related; it's masked log_softmax (obviously related) and nn.CrossEntropyLoss (which calls log_softmax internally). no idea why though; just rereviewed the diff with the benefit of phabricator moved code highlighting

File: aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    const scalar_t* input_data_base,
+    scalar_t* output_data_base,
+    int64_t lastdim_size,
+    int64_t chunk_size,
+    int64_t begin,
+    int64_t end) {
+  if (end <= begin) {
+    return;
+  }
+  using Vec = vec::Vectorized<vec::vec_scalar_t<scalar_t>>;
```
Comment (janeyx99) [inline aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h:41]: What's the diff between this and what I think this sbould be in line 894?
```
using Vec = vec::Vectorized<scalar_t>;
```

This looks similar enough, but just making sure

File: aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+template <typename scalar_t>
+std::pair<int64_t,int64_t> vec_logsoftmax_chunk_size_and_num_chunks(int64_t inner_size, int64_t dim_size) {
+  using Vec = vec::Vectorized<scalar_t>;
+  int64_t BLOCK_SIZE = 128 * 1024;
+  int64_t MAX_CHUNK_SIZE = std::max<int64_t>(BLOCK_SIZE / dim_size / sizeof(scalar_t), Vec::size());
+  MAX_CHUNK_SIZE = MAX_CHUNK_SIZE / Vec::size() * Vec::size();
+  int64_t CHUNK_SIZE = std::min<int64_t>(MAX_CHUNK_SIZE, inner_size);
+  int64_t num_chunks = (inner_size + CHUNK_SIZE - 1) / CHUNK_SIZE;
```
Comment (janeyx99) [inline aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h]: perf optimization instead of using divup?

Comment (swolchok) [reply aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h]: I can't use divup because it's in ATen/Parallel.h or ATen/Parallel-inl.h, which aren't standalone.

File: aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h (hunk 3 above)
Comment (swolchok) [reply aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h:41]: This is meant to be coming from the old line 39 of SoftMaxKernel.cpp, which is the first line of _vec_log_softmax_lastdim.

**Comment by swolchok** [comment]
failures here are hardware-specific; tests are failing on graviton 4 (m8g) but not graviton 3 (m7g). this is odd because they are both supposed to be SVE256.

File: aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h (hunk 4 above)
Comment (malfet) [reply aten/src/ATen/native/cpu/LogSoftmaxKernelImpl.h]: Can you just re-implement `div_up` in the header (as static inline) and add todo to merge it?

**Comment by swolchok** [comment]
Huge props to @malfet for solving this one. I forgot `inline namespace CPU_CAPABILITY` in my extracted header (it wasn't and didn't need to be in the source file because everything was in an anonymous namespace). m8g instances (Graviton 4) have 128-bit SVE support and we only support SVE256. What happened was the SVE256 code incorrectly ran anyway on m8g, which of course silently ran it with 128-bit vector width, hence output item number 5 (i.e., one more than the 4-float vector width of 128-bit SVE) being uninitialized.

**Comment by swolchok** [comment]
converting to draft because 1) I want the existing aarch64 test run to finish but 2) I want to make the block size in `vec_logsoftmax_chunk_size_and_num_chunks` programmable; bc might make changing that later clunky
